frontend/src/pulldasher/filtered-pulls-state.ts

```diff
@@ -2,24 +2,31 @@ import { useState } from 'react';
 import { Pull } from  '../pull';
 
 export type FilterFunction = (pull: Pull) => boolean;
-export type FilterFunctionSetter = (filter: FilterFunction|null) => void;
-export interface FilterProps {
-   filter: FilterFunction
-}
-
+export type FilterFunctionSetter = (filterName:string, filter:FilterFunction|null) => void;
+// Map of human name to filter functions
+type Filters = Record<string, FilterFunction>;
 type ReturnType = [Pull[], FilterFunctionSetter];
 
-const defaultFilter = (pull: Pull) => !!pull;
-
 /**
- * Wrapper around an array of pulls that provides filtering
+ * Wrapper around an array of pulls that allows multiple filters to be
+ * specified.
+ * setFilter(name, func) will *remove* the filter if func is null
  */
 export function useFilteredPullsState(pulls: Pull[]): ReturnType {
-   const [{filter}, setFilter] = useState<FilterProps>({filter: defaultFilter});
+   const [filters, setFilter] = useState<Filters>({});
    return [
-      pulls.filter(filter),
-      (filter: FilterFunction) => {
-         setTimeout(() => setFilter({filter: filter || defaultFilter}), 0);
+      filterPulls(pulls, filters),
+      (filterName:string, filter:FilterFunction|null) => {
+         if (filter) {
+            filters[filterName] = filter;
+         } else {
+            delete filters[filterName];
+         }
+         setTimeout(() => setFilter({...filters}), 0);
```

Comment:
This seems ... odd, why are we waiting until we are finished to call `setFilter`? It's a code smell to me. I see you are keeping what was working, but it's worth a poke.

I think you might be able to do something like:

`setFilter(() => {...filters})` to defer it with react

Reply:
Commit c0adbe4858aaf3866c5a5b3d93fdb43d30fb2273 tells the story here. using `setFitler(() => ...)` was my first instinct too, but it still resulted in it happening in the same callstack as the keyboard event and holding up the main UI thread.  But... I'll try it again in case I got it wrong.

Reply:
Nope, `setState(() => something)` is functional, but it *doesn't* defer the code execution. From my reading of the docs, it's mainly used for when your new state is predictably derived from the previous state: `(bool) => !bool` or something; the docs don't mention it running that function deferred.

We could of course have the *caller* do this, but that felt more like a gotcha.

Reply:
Weird something is certainly off setstate is async: https://javascript.plainenglish.io/react-tips-async-and-setstate-cb539ad62135

Reply:
Here's profiling erasing a search of `z` .. i.e. one backspace keydown:
![image](https://user-images.githubusercontent.com/26855/159425019-5e4a9f64-813c-4806-ac98-3383a7f0ce09.png)

Without the setTimeout, the whole render takes place in the event handler:
![image](https://user-images.githubusercontent.com/26855/159425255-828fbcc1-1843-4069-9f0e-e92667070ddc.png)

Reply:
So, no, `setState()` is not async :shrug: ..   maybe if you call it multiple times? ..... no, not then either.

Reply:
Huh super weird. thanks for all the time you spent digging.